src/providers/message-provider.tsx

```diff
@@ -0,0 +1,472 @@
+          Math.floor((since || Date.now()) / 1000)
+        );
+        setSince(Date.now());
+      },
+      since === 0 ? 500 : 10000
+    );
+
+    return () => {
+      clearTimeout(timer);
+    };
+  }, [since, messageServiceReady, messageService, chat.channels]);
+
+  // // Ready state handler
+  const handleReadyState = () => {
+    setMessageServiceReady(true);
+  };
+
+  useEffect(() => {
+    messageService?.removeListener(MessageEvents.Ready, handleReadyState);
+    messageService?.addListener(MessageEvents.Ready, handleReadyState);
+
+    return () => {
+      messageService?.removeListener(MessageEvents.Ready, handleReadyState);
+    };
+  }, [messageServiceReady, messageService]);
+
+  // Profile update handler
+  const handleProfileUpdate = (data: Profile[]) => {
+    props.addProfile(data);
+  };
+
+  useEffect(() => {
+    messageService?.removeListener(MessageEvents.ProfileUpdate, handleProfileUpdate);
+    messageService?.addListener(MessageEvents.ProfileUpdate, handleProfileUpdate);
+    return () => {
+      messageService?.removeListener(MessageEvents.ProfileUpdate, handleProfileUpdate);
+    };
+  }, [messageService, chat.profiles]);
+
+  //Direct contact handler
+  const handleDirectContact = (data: DirectContact[]) => {
+    const result = [...chat.directContacts];
+    data.forEach(({ name, pubkey }) => {
+      const isPresent = chat.directContacts.some(
+        (obj) => obj.name === name && obj.pubkey === pubkey
+      );
+      if (!isPresent) {
+        result.push({ name, pubkey });
+      }
+    });
+    if (result.length !== 0) {
+      props.addDirectContacts(result);
+    }
+  };
+
+  useEffect(() => {
+    messageService?.removeListener(MessageEvents.DirectContact, handleDirectContact);
+    messageService?.addListener(MessageEvents.DirectContact, handleDirectContact);
+    return () => {
+      messageService?.removeListener(MessageEvents.DirectContact, handleDirectContact);
+    };
+  }, [messageService]);
+
+  //Direct message ahandle before sent
+
+  const handleDirectMessageBeforeSent = (data: DirectMessage[]) => {
+    data.map((m) => {
+      const { peer, id } = m;
+      setReplacedDirectMessagesBuffer((prevBuffer) => [...prevBuffer, id]);
+      props.addDirectMessages(peer, m);
+      checkDirectMessageSending(peer, m);
+    });
+  };
+
+  useEffect(() => {
+    messageService?.removeListener(
+      MessageEvents.DirectMessageBeforeSent,
+      handleDirectMessageBeforeSent
+    );
+    messageService?.addListener(
+      MessageEvents.DirectMessageBeforeSent,
+      handleDirectMessageBeforeSent
+    );
+
+    return () => {
+      messageService?.removeListener(
+        MessageEvents.DirectMessageBeforeSent,
+        handleDirectMessageBeforeSent
+      );
+    };
+  }, [messageService, chat.directMessages]);
+
+  const checkDirectMessageSending = (peer: string, data: DirectMessage) => {
+    setTimeout(() => {
+      props.verifyDirectMessageSending(peer, data);
+    }, 20000);
```

Comment:
I have tried to use this hook. But since it is a hook it does not work inside the functions. But it will work in a React Functional Component. In our case we want to use setTimeOut inside the checkDirectMessageSending function.